include/nbl/video/IGPUCommandPool.h

```diff
@@ -31,13 +66,219 @@ class IGPUCommandPool : public core::IReferenceCounted, public IBackendObject
+        if (address == decltype(params.m_commandAllocator)::invalid_address)
+            return nullptr;
+
+        wipeNextCommandSize();
+
+        void* cmdMem = m_data + address;
+        return new (cmdMem) Cmd(args...);
+    }
+
+private:
+    void wipeNextCommandSize()
+    {
+        const auto cursor = params.m_commandAllocator.get_allocated_size();
+        const uint32_t wipeSize = offsetof(IGPUCommandPool::ICommand, m_size) + sizeof(IGPUCommandPool::ICommand::m_size);
+        if (cursor + wipeSize < params.m_commandAllocator.get_total_size())
+            memset(m_data + cursor, 0, wipeSize);
+    }
+};
+
+struct IGPUCommandPool::CommandSegment::Iterator
+{
+    CommandSegment* m_segment = nullptr;
+    ICommand* m_cmd = nullptr;
+};
+
+class IGPUCommandPool::CBindIndexBufferCmd : public IGPUCommandPool::ICommand
+{
+public:
+    CBindIndexBufferCmd(const core::smart_refctd_ptr<const video::IGPUBuffer>& indexBuffer) : ICommand(calc_size(indexBuffer)), m_indexBuffer(indexBuffer) {}
+
+    static uint32_t calc_size(const core::smart_refctd_ptr<const video::IGPUBuffer>& indexBuffer)
+    {
+        return core::alignUp(sizeof(void*), alignof(CBindIndexBufferCmd));
+    }
+
+protected:
+    core::smart_refctd_ptr<const video::IGPUBuffer> m_indexBuffer;
+};
+
+class IGPUCommandPool::CDrawCmd : public IGPUCommandPool::ICommand
+{
+public:
+    CDrawCmd() : ICommand(calc_size()) {}
+
+    static uint32_t calc_size()
+    {
+        return core::alignUp(1u, alignof(CDrawCmd));
+    }
+};
+
+class IGPUCommandPool::CDrawIndexedCmd : public IGPUCommandPool::ICommand
+{
+public:
+    CDrawIndexedCmd() : ICommand(calc_size()) {}
+
+    static uint32_t calc_size()
+    {
+        return core::alignUp(1u, alignof(CDrawIndexedCmd));
+    }
+};
+
+class IGPUCommandPool::CDrawIndirectCommonBase : public IGPUCommandPool::ICommand
+{
+public:
+    CDrawIndirectCommonBase(const core::smart_refctd_ptr<const video::IGPUBuffer>& buffer, const uint32_t size) : ICommand(size), m_buffer(buffer) {}
+
+protected:
+    core::smart_refctd_ptr<const IGPUBuffer> m_buffer;
+};
+
+class IGPUCommandPool::CDrawIndirectCmd : public IGPUCommandPool::CDrawIndirectCommonBase
+{
+public:
+    CDrawIndirectCmd(const core::smart_refctd_ptr<const video::IGPUBuffer>& buffer) : CDrawIndirectCommonBase(buffer, calc_size(buffer)) {}
+
+    static uint32_t calc_size(const core::smart_refctd_ptr<const video::IGPUBuffer>& buffer)
+    {
+        return core::alignUp(sizeof(void*), alignof(CDrawIndirectCmd));
+    }
+};
+
+class IGPUCommandPool::CDrawIndexedIndirectCmd : public IGPUCommandPool::CDrawIndirectCommonBase
+{
+public:
+    CDrawIndexedIndirectCmd(const core::smart_refctd_ptr<const video::IGPUBuffer>& buffer) : CDrawIndirectCommonBase(buffer, calc_size(buffer)) {}
+
+    static uint32_t calc_size(const core::smart_refctd_ptr<const video::IGPUBuffer>& buffer)
+    {
+        return core::alignUp(sizeof(void*), alignof(CDrawIndexedIndirectCmd));
+    }
+};
```

Comment:
you can declare `calc_size` in the common base and probably just return `sizeof(*this)`

Reply:
Can't do because `calc_size` is defined `static`. Anyway I don't think it matters now.

Reply:
yep doesn't matter cause of the CRTP QoL idea